# HG changeset patch
# Parent 35aa6f97ddf2e68fd32aa7e7c9910a6070725ca1

Modify the prior cc-restrictions patch to allow a decrease_reservation memory
operation but clear the memory such that no guest data leaks. This is only
done for non-privileged guests as it is assumed that privileged guests will
be more careful.
This modificaton is necessary to be able to use unmodified Windows PV drivers
which use the decrease_reservation operation to create memory 'holes' to use
for shared_info, grant tables, and the like.

diff -r 35aa6f97ddf2 xen/common/memory.c
--- a/xen/common/memory.c
+++ b/xen/common/memory.c
@@ -205,6 +205,13 @@ int guest_remove_page(struct domain *d, 
 
     guest_physmap_remove_page(d, gmfn, mfn, 0);
 
+    if ( unlikely(opt_cc_restrictions) && !IS_PRIV(d) )
+    {
+        char *b = map_domain_page(page_to_mfn(page));
+        clear_page(b);
+        unmap_domain_page(b);
+    }
+
     put_page(page);
 
     return 1;
@@ -519,7 +526,7 @@ long do_memory_op(unsigned long cmd, XEN
      * return memory to Xen. */
     if ( unlikely(opt_cc_restrictions) &&
          !IS_PRIV(current->domain) &&
-         (op == XENMEM_decrease_reservation || op == XENMEM_exchange) )
+         ( op == XENMEM_exchange) )
         return -EPERM;
 
     switch ( op )
